lib/gc/Analysis/GlobalAnalysis.cpp

```diff
@@ -0,0 +1,651 @@
+  // whether to push all new axis at the beginning of outer perm
+  targetOuterAxis.insert(targetOuterAxis.begin(), newDimAxis.begin(),
+                         newDimAxis.end());
+  for (auto &&[ia, ts] : llvm::zip(baseInnerAxis, baseTileSizes)) {
+    if (reverseIndexMap[ia] >= 0) {
+      targetInnerAxis.push_back(reverseIndexMap[ia]);
+      targetTileSizes.push_back(ts);
+    }
+  }
+  return TensorLayout(targetOuterAxis, targetInnerAxis, targetTileSizes);
+}
+
+// TODO(yifei): enhance the logic for choose base input index
+static size_t getBaseInputIdx(ArrayRef<TensorLayout> curInputLayouts) {
+  for (size_t i = 0; i < curInputLayouts.size(); ++i) {
+    if (!curInputLayouts[i].isPlain()) {
+      return i;
+    }
+  }
+  return 0;
+}
+
+std::pair<SmallVector<int64_t>, SmallVector<int64_t>>
+getPackingAxis(int64_t numRank, bool transposed) {
+  assert(numRank >= 2 &&
+         "The rank of matmul semantic contraction op shall be at least 2.");
+  SmallVector<int64_t> outerAxisPerm(numRank);
+  SmallVector<int64_t> innerAxisPos(2);
+  std::iota(outerAxisPerm.begin(), outerAxisPerm.end(), 0);
+  innerAxisPos[0] = numRank - 2;
+  innerAxisPos[1] = numRank - 1;
+  if (transposed) {
+    std::swap(outerAxisPerm[numRank - 2], outerAxisPerm[numRank - 1]);
+    std::swap(innerAxisPos[0], innerAxisPos[1]);
+  }
+  return std::make_pair(outerAxisPerm, innerAxisPos);
+}
+
+// copied from mlir
+static SmallVector<int64_t>
+projectToInnerMostNonUnitDimsPos(ArrayRef<int64_t> dimsPos,
+                                 ArrayRef<ReassociationIndices> reassocIndices,
+                                 ArrayRef<int64_t> targetShape) {
+  SmallVector<int64_t> projectedDimsPos;
+  for (auto pos : dimsPos) {
+    // In the case all dims are unit, this will return the inner-most one.
+    int64_t projectedPos = reassocIndices[pos].back();
+    for (auto i : llvm::reverse(reassocIndices[pos])) {
+      int64_t dim = targetShape[i];
+      if (dim > 1 || ShapedType::isDynamic(dim)) {
+        projectedPos = i;
+        break;
+      }
+    }
+    projectedDimsPos.push_back(projectedPos);
+  }
+  return projectedDimsPos;
+}
+
+// Check if all dims in dimsPos are divisible by the corresponding tile sizes.
+static bool isDimsDivisibleByTileSizes(ArrayRef<int64_t> dimsPos,
+                                       ArrayRef<int64_t> shape,
+                                       ArrayRef<int64_t> tileSizes) {
+  return llvm::all_of(llvm::zip_equal(dimsPos, tileSizes),
+                      [shape](std::tuple<int64_t, int64_t> sizePair) {
+                        int64_t dim = shape[std::get<0>(sizePair)];
+                        return !ShapedType::isDynamic(dim) &&
+                               (dim % std::get<1>(sizePair)) == 0;
+                      });
+}
+
+// if forceBlocking is set to true, we will unconditionally convert
+// input/weight/output to blocking layout; otherwise we follow the default
+// heuristic logic
```

Comment:
```suggestion
// if forceBlocking is set to true, we will unconditionally convert
// input/weight/output to blocking layout; otherwise we follow the default
// heuristic logic, which is ?...
```